### PR TITLE
Fixes #420 Decorate exported class

### LINKED ISSUE #420: Cannot directly export classes with decorators
Playground:

- [`@foo class Bar`](https://civet.dev/playground/?code=aW1wb3J0IHsgaHRtbCwgY3NzLCBMaXRFbGVtZW50IH0gZnJvbSAnbGl0JwppbXBvcnQgeyBjdXN0b21FbGVtZW50LCBwcm9wZXJ0eSB9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJwoKQEBjdXN0b21FbGVtZW50KCdzaW1wbGUtZ3JlZXRpbmcnKQpjbGFzcyBTaW1wbGVHcmVldGluZyBleHRlbmRzIExpdEVsZW1lbnQKICBAc3R5bGVzID0gY3NzYHAgeyBjb2xvcjogYmx1ZSB9YAoKICBAQHByb3BlcnR5KCkKICBuYW1lID0gJ1NvbWVib2R5JwoKICByZW5kZXIoKQogICAgaHRtbGA8cD5IZWxsbywgJHt0aGlzLm5hbWV9ITwvcD5gCg%3D%3D)
- [`@foo export class Bar`](https://civet.dev/playground/?code=aW1wb3J0IHsgaHRtbCwgY3NzLCBMaXRFbGVtZW50IH0gZnJvbSAnbGl0JwppbXBvcnQgeyBjdXN0b21FbGVtZW50LCBwcm9wZXJ0eSB9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJwoKQEBjdXN0b21FbGVtZW50KCdzaW1wbGUtZ3JlZXRpbmcnKQpleHBvcnQgY2xhc3MgU2ltcGxlR3JlZXRpbmcgZXh0ZW5kcyBMaXRFbGVtZW50CiAgQHN0eWxlcyA9IGNzc2BwIHsgY29sb3I6IGJsdWUgfWAKCiAgQEBwcm9wZXJ0eSgpCiAgbmFtZSA9ICdTb21lYm9keScKCiAgcmVuZGVyKCkKICAgIGh0bWxgPHA%2BSGVsbG8sICR7dGhpcy5uYW1lfSE8L3A%2BYAo%3D) (ParseError)
- [`export @foo class Bar`](https://civet.dev/playground/?code=aW1wb3J0IHsgaHRtbCwgY3NzLCBMaXRFbGVtZW50IH0gZnJvbSAnbGl0JwppbXBvcnQgeyBjdXN0b21FbGVtZW50LCBwcm9wZXJ0eSB9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJwoKZXhwb3J0IEBAY3VzdG9tRWxlbWVudCgnc2ltcGxlLWdyZWV0aW5nJykKY2xhc3MgU2ltcGxlR3JlZXRpbmcgZXh0ZW5kcyBMaXRFbGVtZW50CiAgQHN0eWxlcyA9IGNzc2BwIHsgY29sb3I6IGJsdWUgfWAKCiAgQEBwcm9wZXJ0eSgpCiAgbmFtZSA9ICdTb21lYm9keScKCiAgcmVuZGVyKCkKICAgIGh0bWxgPHA%2BSGVsbG8sICR7dGhpcy5uYW1lfSE8L3A%2BYAo%3D) (Incompatible with TypeScript)

### PR DESCRIPTION
**Pull request by STRd6**
This allows decorators in TS "before export" and TC39 "after export" positions and doesn't try to do anything smart about it yet.

I added another state flag for when we are inside `Decorators` to prevent `@@decorator class ...` from being treated as an implicit call passing a class expression to the decorator function. If someone really wants to do that they'll need to use parentheses and be explicit.

### CODE REVIEW

File: source/parser.hera [hunk 1]
```diff
@@ -5349,6 +5349,9 @@ TypeDeclaration
   ( Export _? )? ( Declare _? )  TypeLexicalDeclaration -> { ts: true, children: $0 }
   ( Export _? )? ( Declare _? )? TypeDeclarationRest    -> { ts: true, children: $0 }
 
+  # NOTE: TS Exported classes can have decorators before export but not after
+  Decorators ( Export _? ) !Decorators ClassDeclaration -> { ts: true, children: $0 }
```
Comment (edemaine) [inline source/parser.hera:5353]: I don't think this should be `ts: true`. Exported classes should be in JS code too, just like other decorated classes.

Related, why is this here and not in `ExportDeclaration`?

Ah, I see, this type of decorating is inconsistent with the [TC39 decorator proposal](https://github.com/tc39/proposal-decorators)...

**Comment by edemaine** [comment]
Could you explain (maybe with a test case) why we need another flag here? Oh, was this a consequence of extending to CallExpression in a previous PR? Kind of surprised any tests passed...

**Review by edemaine — APPROVED** [review]

**Comment by STRd6** [comment]
Yeah, the idea was to keep allowing `fn class ...` as an implicit application generally but disallow it in inline decorators.
